File: Arrays/217_ContainsDuplicate.js
/**
 * ╔══════════════════════════════════════════════════════════════╗
 * ║  LeetCode #217 — Contains Duplicate                         ║
 * ║  Link       : https://leetcode.com/problems/contains-duplicate/
 * ║  Difficulty : Easy                                          ║
 * ║  Topic      : Arrays, Hashing                               ║
 * ╚══════════════════════════════════════════════════════════════╝
 *
 * PROBLEM STATEMENT:
 * ──────────────────
 * Given an integer array `nums`, return true if any value appears
 * at least twice in the array, and return false if every element
 * is distinct.
 *
 * EXAMPLES:
 * ─────────
 * Input  : nums = [1, 2, 3, 1]  →  Output : true   (1 appears twice)
 * Input  : nums = [1, 2, 3, 4]  →  Output : false  (all unique)
 * Input  : nums = [1,1,1,3,3,4,3,2,4,2]  →  Output : true
 *
 * Constraints:
 *   • 1 <= nums.length <= 10^5
 *   • -10^9 <= nums[i] <= 10^9
 */


// ══════════════════════════════════════════════════════════════
// APPROACH 1 — Brute Force (Nested Loops)
// ══════════════════════════════════════════════════════════════
// Idea  : Compare every pair (i, j) where i < j.
//         If nums[i] === nums[j], a duplicate exists.
//
// Time  : O(n²) — two nested loops
// Space : O(1)  — no extra memory
//
// ❌ Too slow for large inputs. TLE on LeetCode.

function bruteForce(nums) {
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {   // only pairs ahead of i
      if (nums[i] === nums[j]) return true;         // duplicate found
    }
  }
  return false;
}


// ══════════════════════════════════════════════════════════════
// APPROACH 2 — Sorting
// ══════════════════════════════════════════════════════════════
// Idea  : Sort the array so duplicates become adjacent neighbors.
//         One pass is enough to spot nums[i] === nums[i+1].
//
// Time  : O(n log n) — dominated by sort
// Space : O(1)       — in-place sort (JS sort mutates the array)
//
// ⚠️  Mutates the original array.

function containsDuplicateSort(nums) {
  nums.sort((a, b) => a - b);                      // numeric sort (not lexicographic!)
  for (let i = 0; i < nums.length - 1; i++) {
    if (nums[i] === nums[i + 1]) return true;       // neighbors match → duplicate
  }
  return false;
}


// ══════════════════════════════════════════════════════════════
// APPROACH 3 — Hash Set  (Optimal ✅)
// ══════════════════════════════════════════════════════════════
// Idea  : Use a Set to remember numbers already seen.
//         - If current number IS in the set → duplicate found.
//         - If NOT → add it and keep going.
//
// Why a Set?
//   JavaScript's Set provides O(1) average time for .has() and
//   .add(), making the overall pass O(n).
//
// Time  : O(n)  — single pass
// Space : O(n)  — set holds at most n unique values

/**
 * @param {number[]} nums
 * @return {boolean}
 */
var containsDuplicate = function (nums) {

  // Step 1: Create an empty Set to track seen numbers
  const seen = new Set();

  // Step 2: Iterate through every number in the array
  for (const num of nums) {

    // Step 3: Check if we've seen this number before
    if (seen.has(num)) {
      return true;          // duplicate found → exit early
    }

    // Step 4: Mark this number as seen
    seen.add(num);
  }

  // Step 5: Finished the loop with no duplicates found
  return false;
};


// ══════════════════════════════════════════════════════════════
// ONE-LINER (JavaScript Bonus)
// ══════════════════════════════════════════════════════════════
// A Set only stores unique values. If the Set's size is less
// than the array's length, there must be at least one duplicate.
//
// Time  : O(n)  |  Space : O(n)
// ⚠️  Always builds the full Set — no early exit.

const oneLiner = (nums) => new Set(nums).size < nums.length;


// ══════════════════════════════════════════════════════════════
// TESTS
// ══════════════════════════════════════════════════════════════
const assert = (condition, msg) => {
  if (!condition) throw new Error(`FAILED: ${msg}`);
};

assert(containsDuplicate([1, 2, 3, 1])              === true,  "Test 1");
assert(containsDuplicate([1, 2, 3, 4])              === false, "Test 2");
assert(containsDuplicate([1, 1, 1, 3, 3, 4, 3, 2]) === true,  "Test 3");
assert(containsDuplicate([1])                        === false, "Test 4");

console.log("All tests passed [OK]");


/**
 * ╔══════════════════════════════════════════════════════════════╗
 * ║  COMPLEXITY SUMMARY                                         ║
 * ╠══════════════════════════════════════════════════════════════╣
 * ║  Approach          Time        Space   Notes                ║
 * ║  ─────────────── ──────────── ─────── ──────────────────── ║
 * ║  Brute Force      O(n²)        O(1)   TLE on large input   ║
 * ║  Sorting          O(n log n)   O(1)   Mutates array        ║
 * ║  Hash Set (Set)   O(n)         O(n)   Best general-purpose ║
 * ║  One-liner        O(n)         O(n)   No early exit        ║
 * ╚══════════════════════════════════════════════════════════════╝
 *
 * KEY TAKEAWAYS:
 * ──────────────
 * • JS `Set` is the equivalent of Python's `set` — use it for
 *   "have I seen this before?" problems.
 *
 * • Always use `.sort((a, b) => a - b)` for number arrays.
 *   Default JS sort is LEXICOGRAPHIC — [10, 2, 1] would sort as
 *   [1, 10, 2] which gives WRONG results for this problem.
 *
 * • Early return (return true inside loop) is more efficient than
 *   the one-liner because it stops as soon as the first duplicate
 *   is found instead of scanning the whole array.
 *
 * • This Set pattern is the foundation for:
 *     → Two Sum (#1)
 *     → Longest Consecutive Sequence (#128)
 *     → Happy Number (#202)
 */

File: Arrays/217_ContainsDuplicate.py
"""
╔══════════════════════════════════════════════════════════════╗
║  LeetCode #217 — Contains Duplicate                         ║
║  Link       : https://leetcode.com/problems/contains-duplicate/
║  Difficulty : Easy                                          ║
║  Topic      : Arrays, Hashing                               ║
╚══════════════════════════════════════════════════════════════╝

PROBLEM STATEMENT:
──────────────────
Given an integer array `nums`, return True if any value appears
at least twice in the array, and return False if every element
is distinct.

EXAMPLES:
─────────
Example 1:
  Input  : nums = [1, 2, 3, 1]
  Output : True   ← 1 appears twice

Example 2:
  Input  : nums = [1, 2, 3, 4]
  Output : False  ← all elements are unique

Example 3:
  Input  : nums = [1, 1, 1, 3, 3, 4, 3, 2, 4, 2]
  Output : True

Constraints:
  • 1 <= nums.length <= 10⁵
  • -10⁹ <= nums[i] <= 10⁹
"""

# ══════════════════════════════════════════════════════════════
# APPROACH 1 — Brute Force (Nested Loops)
# ══════════════════════════════════════════════════════════════
# Idea  : Compare every pair (i, j) where i < j.
#         If nums[i] == nums[j], a duplicate exists.
#
# Time  : O(n²) — two nested loops over the array
# Space : O(1)  — no extra memory used
#
# ❌ Too slow for large inputs. Only good to understand the problem.

def brute_force(nums: list[int]) -> bool:
    n = len(nums)
    for i in range(n):
        for j in range(i + 1, n):          # only check pairs ahead of i
            if nums[i] == nums[j]:
                return True                 # found a duplicate
    return False


# ══════════════════════════════════════════════════════════════
# APPROACH 2 — Sorting
# ══════════════════════════════════════════════════════════════
# Idea  : After sorting, duplicates will always be ADJACENT.
#         Walk through once and check if nums[i] == nums[i+1].
#
# Time  : O(n log n) — dominated by the sort step
# Space : O(1)       — sort is in-place (ignoring sort's call stack)
#
# ⚠️  This modifies the original array. Use a copy if that matters.

def contains_duplicate_sort(nums: list[int]) -> bool:
    nums.sort()                             # bring duplicates next to each other
    for i in range(len(nums) - 1):
        if nums[i] == nums[i + 1]:         # adjacent elements are equal → duplicate
            return True
    return False


# ══════════════════════════════════════════════════════════════
# APPROACH 3 — Hash Set (Optimal ✅)
# ══════════════════════════════════════════════════════════════
# Idea  : Use a set to remember numbers we've already seen.
#         - If current number IS already in the set → duplicate found.
#         - If NOT → add it to the set and continue.
#
# Why a set?
#   A Python set gives O(1) average-time for both lookup and insert,
#   making the full pass O(n) overall.
#
# Time  : O(n)  — single pass through the array
# Space : O(n)  — set stores at most n unique elements

class Solution:
    def containsDuplicate(self, nums: list[int]) -> bool:

        # ── Step 1: Create an empty set to track seen numbers ──
        seen = set()

        # ── Step 2: Iterate through every number in the array ──
        for num in nums:

            # ── Step 3: Check if we've seen this number before ──
            if num in seen:
                return True             # duplicate found → stop early

            # ── Step 4: Mark this number as seen ───────────────
            seen.add(num)

        # ── Step 5: Looped through everything with no duplicate ─
        return False


# ══════════════════════════════════════════════════════════════
# ONE-LINER (Pythonic Bonus 🐍)
# ══════════════════════════════════════════════════════════════
# A set only keeps unique values. If the set is smaller than
# the original list, there must be a duplicate.
#
# Time  : O(n)  |  Space : O(n)

def one_liner(nums: list[int]) -> bool:
    return len(set(nums)) < len(nums)


# ══════════════════════════════════════════════════════════════
# DRIVER / TESTS
# ══════════════════════════════════════════════════════════════
if __name__ == "__main__":
    sol = Solution()

    assert sol.containsDuplicate([1, 2, 3, 1])              == True,  "Test 1 failed"
    assert sol.containsDuplicate([1, 2, 3, 4])              == False, "Test 2 failed"
    assert sol.containsDuplicate([1, 1, 1, 3, 3, 4, 3, 2]) == True,  "Test 3 failed"
    assert sol.containsDuplicate([1])                        == False, "Test 4 failed"

    print("All tests passed [OK]")


"""
╔══════════════════════════════════════════════════════════════╗
║  COMPLEXITY SUMMARY                                         ║
╠══════════════════════════════════════════════════════════════╣
║  Approach          Time        Space   Notes                ║
║  ─────────────── ──────────── ─────── ──────────────────── ║
║  Brute Force      O(n²)        O(1)   Fails large inputs   ║
║  Sorting          O(n log n)   O(1)   Mutates array        ║
║  Hash Set ✅      O(n)         O(n)   Best general-purpose ║
║  One-liner ✅     O(n)         O(n)   Pythonic shorthand   ║
╚══════════════════════════════════════════════════════════════╝

KEY TAKEAWAYS:
──────────────
• A set is your go-to for "have I seen this before?" problems.
• Early return (return True inside the loop) avoids scanning the
  entire array once a duplicate is found → faster in practice.
• The one-liner is clean but doesn't short-circuit early; it always
  builds the full set before comparing lengths.
• This exact pattern (seen = set()) is the foundation for problems
  like Two Sum, Longest Consecutive Sequence, and Group Anagrams.
"""
